src/Adapter/Product/Update/Filler/SeoFiller.php

```diff
@@ -0,0 +1,166 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product\Update\Filler;
+
+use PrestaShop\PrestaShop\Adapter\Category\Repository\CategoryRepository;
+use PrestaShop\PrestaShop\Adapter\Product\Repository\ProductRepository;
+use PrestaShop\PrestaShop\Adapter\Tools;
+use PrestaShop\PrestaShop\Core\Domain\Category\Exception\CategoryNotFoundException;
+use PrestaShop\PrestaShop\Core\Domain\Category\ValueObject\CategoryId;
+use PrestaShop\PrestaShop\Core\Domain\Product\Command\UpdateProductCommand;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\RedirectOption;
+use PrestaShop\PrestaShop\Core\Exception\CoreException;
+use Product;
+
+class SeoFiller implements ProductFillerInterface
+{
+    /**
+     * @var ProductRepository
+     */
+    private $productRepository;
+
+    /**
+     * @var CategoryRepository
+     */
+    private $categoryRepository;
+
+    /**
+     * @var Tools
+     */
+    private $tools;
+
+    /**
+     * @param ProductRepository $productRepository
+     * @param CategoryRepository $categoryRepository
+     * @param Tools $tools
+     */
+    public function __construct(
+        ProductRepository $productRepository,
+        CategoryRepository $categoryRepository,
+        Tools $tools
+    ) {
+        $this->productRepository = $productRepository;
+        $this->categoryRepository = $categoryRepository;
+        $this->tools = $tools;
```

Comment:
It is possible to create a stringutils class instead of injecting legacy tools function for this.


I created an example if you want to use it (did not have time to make proper DI on this) 


```
<?php

namespace PrestaShop\PrestaShop\Core\String;

use Transliterator;

class StringUtils
{
    /**
     * @var bool
     */
    private $allowAccents;

    /**
     * @var Transliterator
     */
    private $transliterator;

    public function __construct(bool $allowAccents)
    {
        $this->allowAccents = $allowAccents;
        $this->transliterator = Transliterator::create('Any-Latin; Latin-ASCII');
    }

    public function str2url($str)
    {
        static $array_str = [];

        if (!is_string($str)) {
            return false;
        }

        if (isset($array_str[$str])) {
            return $array_str[$str];
        }

        if ($str == '') {
            return '';
        }

        $return_str = trim($str);
        $return_str = mb_strtolower($return_str, 'UTF-8');

        if (!$this->allowAccents) {
            $return_str = $this->transliterator->transliterate($return_str);
        }

        // Remove all non-whitelist chars.
        if ($this->allowAccents) {
            $return_str = preg_replace('/[^a-zA-Z0-9\s\'\:\/\[\]\-\p{L}]/u', '', $return_str);
        } else {
            $return_str = preg_replace('/[^a-zA-Z0-9\s\'\:\/\[\]\-]/', '', $return_str);
        }

        $return_str = preg_replace('/[\s\'\:\/\[\]\-]+/', ' ', $return_str);
        $return_str = str_replace([' ', '/'], '-', $return_str);

        $array_str[$str] = $return_str;

        return $return_str;
    }
}

```

Reply:
I agree it would be nice!! Although it's not exactly the scope of this PR so I don't think it should block it It can be done in a dedicated PR that would replace this usage everywhere in the codebase

Reply:
For this can of function, I even think maybe a static function could be nice, the injection, in this case, is not that relevant

I'm usually very much against the usage of static methods (especially the way they've been abused in ObjectModel) but in this kind of case it's quite adapted actually since the method is completely stateless

Reply:
Actually it depends on the configuration :)

Agree this is non blocking